File: apps/api/app/routers/group_settlements.py
from fastapi import APIRouter, status

from app.deps.auth import CurrentUser
from app.deps.currency import DisplayCurrency
from app.deps.db import SessionDep
from app.schemas.group_settlement import (
    GroupBalancesResponse,
    GroupMoneySettingsResponse,
    GroupMoneySettingsUpdate,
    GroupSettlementCreate,
    GroupSettlementLegUpdate,
    GroupSettlementResponse,
    GroupWriteOffCreate,
)
from app.services import group_money_service, group_settlement_service

# The balances half of a group's money, on its own prefix for the same reason the expenses half is:
# the group router stays about the people entity, which is what keeps that entity money-agnostic.
router = APIRouter(prefix="/groups/{group_id}", tags=["shared money"])


# Returns every member's balance per currency, plus the fewest payments that clear each bucket.
# Balances never net across currencies — each is its own settle line — so a member can appear as a
# creditor in one bucket and a debtor in another.
@router.get("/balances", response_model=GroupBalancesResponse)
async def get_group_balances(
    group_id: int,
    current_user: CurrentUser,
    session: SessionDep,
    currency: DisplayCurrency = None,
) -> GroupBalancesResponse:
    return await group_settlement_service.get_balances(session, group_id, current_user, currency=currency)


# Lists the group's recorded settlements and write-offs, newest first.
@router.get("/settlements", response_model=list[GroupSettlementResponse])
async def list_group_settlements(
    group_id: int,
    current_user: CurrentUser,
    session: SessionDep,
) -> list[GroupSettlementResponse]:
    return await group_settlement_service.list_settlements(session, group_id, current_user)


# Records a payment one member made to another. Lands pending unless the group has opted into
# auto-finalise. Returns 400 for a leg that names an amount without an account, or crosses currencies
# without saying what moved; 404 for a seat or account that is not reachable.
@router.post("/settlements", response_model=GroupSettlementResponse, status_code=status.HTTP_201_CREATED)
async def create_group_settlement(
    group_id: int,
    body: GroupSettlementCreate,
    current_user: CurrentUser,
    session: SessionDep,
) -> GroupSettlementResponse:
    return await group_settlement_service.record_settlement(
        session,
        group_id,
        current_user,
        from_member_id=body.from_member_id,
        to_member_id=body.to_member_id,
        date=body.date,
        amount=body.amount,
        currency=body.currency,
        from_account_id=body.from_account_id,
        from_amount=body.from_amount,
        to_account_id=body.to_account_id,
        to_amount=body.to_amount,
        notes=body.notes,
    )


# Records a debt the creditor gives up on. Clears the same bucket a payment would and moves no money.
# Returns 403 when the caller is not the seat being owed.
@router.post("/settlements/write-off", response_model=GroupSettlementResponse, status_code=status.HTTP_201_CREATED)
async def create_group_write_off(
    group_id: int,
    body: GroupWriteOffCreate,
    current_user: CurrentUser,
    session: SessionDep,
) -> GroupSettlementResponse:
    return await group_settlement_service.record_write_off(
        session,
        group_id,
        current_user,
        from_member_id=body.from_member_id,
        to_member_id=body.to_member_id,
        date=body.date,
        amount=body.amount,
        currency=body.currency,
        notes=body.notes,
    )


# Acknowledges receipt of a pending settlement. Only the payee may (403 otherwise); returns 409 when
# it is already confirmed.
@router.post("/settlements/{settlement_id}/confirm", response_model=GroupSettlementResponse)
async def confirm_group_settlement(
    group_id: int,
    settlement_id: int,
    current_user: CurrentUser,
    session: SessionDep,
) -> GroupSettlementResponse:
    return await group_settlement_service.confirm_settlement(session, group_id, settlement_id, current_user)


# Takes a confirmation back, returning the settlement to pending so it can be corrected or removed.
# Only the payee may (403 otherwise).
@router.delete("/settlements/{settlement_id}/confirm", response_model=GroupSettlementResponse)
async def unconfirm_group_settlement(
    group_id: int,
    settlement_id: int,
    current_user: CurrentUser,
    session: SessionDep,
) -> GroupSettlementResponse:
    return await group_settlement_service.unconfirm_settlement(session, group_id, settlement_id, current_user)


# Attaches or clears the caller's OWN cash leg. Which side it lands on follows from which seat they
# hold, so nothing in the body names a side. Returns 400 when the caller is on neither side, and 409
# for a write-off, which moved no money to attach an account to.
@router.put("/settlements/{settlement_id}/account", response_model=GroupSettlementResponse)
async def set_group_settlement_leg(
    group_id: int,
    settlement_id: int,
    body: GroupSettlementLegUpdate,
    current_user: CurrentUser,
    session: SessionDep,
) -> GroupSettlementResponse:
    return await group_settlement_service.set_leg(
        session,
        group_id,
        settlement_id,
        current_user,
        account_id=body.account_id,
        leg_amount=body.amount,
    )


# Removes a settlement — which is what reversing one is. Either named member may while it is pending;
# a write-off only its creditor. A confirmed settlement is refused with 409 until it is un-confirmed.
@router.delete("/settlements/{settlement_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_group_settlement(
    group_id: int,
    settlement_id: int,
    current_user: CurrentUser,
    session: SessionDep,
) -> None:
    await group_settlement_service.delete_settlement(session, group_id, settlement_id, current_user)


# Returns the group's shared money settings.
@router.get("/money-settings", response_model=GroupMoneySettingsResponse)
async def get_group_money_settings(
    group_id: int,
    current_user: CurrentUser,
    session: SessionDep,
) -> GroupMoneySettingsResponse:
    return await group_money_service.get_settings(session, group_id, current_user)


# Updates the group's shared money settings. Admin only (403 otherwise) — setting the standard the
# group holds itself to is management, not money movement.
@router.put("/money-settings", response_model=GroupMoneySettingsResponse)
async def update_group_money_settings(
    group_id: int,
    body: GroupMoneySettingsUpdate,
    current_user: CurrentUser,
    session: SessionDep,
) -> GroupMoneySettingsResponse:
    return await group_money_service.update_settings(
        session,
        group_id,
        current_user,
        default_split_method=body.default_split_method,
        auto_finalise_settlements=body.auto_finalise_settlements,
    )

File: apps/web/lib/api/group-settlements.ts
/*
 * Server-only types + mappers for a group's balances and the settlements that clear them. Reads go
 * through the three functions here; mutations are server actions in
 * `app/(protected)/shared/settlement-actions.ts`.
 *
 * Three shapes of the API's honesty survive into these types unchanged:
 *
 *   * Balances NEVER net across currencies. One bucket per currency, each with its own zero-sum and
 *     its own settle line — owing dollars while being owed pesos is a real state, and merging the two
 *     would invent a rate nobody agreed to. `myConvertedBalance` is a glance figure at the viewer's
 *     own rate and is never what anybody settles.
 *   * `canConfirm` / `canDelete` are resolved SERVER-SIDE, because they follow from who the caller is
 *     and a second copy of them would be a second copy of a permission check. Render from them.
 *   * A settlement carries up to three amounts, and the two cash legs are each set only when that
 *     side crossed currencies. `null` there means "the account moved exactly what came off the
 *     bucket", not "unknown" — there is no stored rate anywhere, the pair of amounts is the record.
 *
 * Money figures stay STRINGS end to end, for the reason every other module keeps them: they are
 * NUMERIC(18,2) server-side and a JS number would round the digits a balance is checked against.
 */

import 'server-only';

import { authenticatedFetch } from '@/lib/authenticated-fetch';
import type { SettlementStatus } from '@/lib/constants/group-settlements';
import type { SplitMethod } from '@/lib/constants/shared-expenses';

// --- Raw types (API JSON shape, snake_case) ---

interface GroupMemberBalanceRaw {
  member_id: number;
  display_name: string;
  amount: string;
  is_self: boolean;
}

interface GroupSettleSuggestionRaw {
  from_member_id: number;
  from_display_name: string;
  to_member_id: number;
  to_display_name: string;
  amount: string;
}

interface GroupCurrencyBalanceRaw {
  currency: string;
  balances: GroupMemberBalanceRaw[];
  suggestions: GroupSettleSuggestionRaw[];
  my_balance: string;
  my_converted_balance: string | null;
}

interface GroupBalancesRaw {
  group_id: number;
  buckets: GroupCurrencyBalanceRaw[];
  display_currency: string | null;
  skipped_currencies: string[];
}

interface GroupSettlementRaw {
  id: number;
  group_id: number;
  from_member_id: number;
  from_display_name: string;
  to_member_id: number;
  to_display_name: string;
  date: string;
  amount: string;
  currency: string;
  status: SettlementStatus;
  from_account_id: number | null;
  from_amount: string | null;
  to_account_id: number | null;
  to_amount: string | null;
  confirmed_at: string | null;
  notes: string | null;
  can_confirm: boolean;
  can_delete: boolean;
  created_at: string;
  updated_at: string;
}

interface GroupMoneySettingsRaw {
  group_id: number;
  default_split_method: SplitMethod;
  auto_finalise_settlements: boolean;
}

// --- Frontend types (camelCase) ---

export interface GroupMemberBalance {
  memberId: number;
  displayName: string;
  // Positive when this member is owed, negative when they owe. A member who is square has no row at
  // all — a line of zeros on every screen says nothing.
  amount: string;
  isSelf: boolean;
}

export interface GroupSettleSuggestion {
  fromMemberId: number;
  fromDisplayName: string;
  toMemberId: number;
  toDisplayName: string;
  amount: string;
}

export interface GroupCurrencyBalance {
  currency: string;
  balances: GroupMemberBalance[];
  // The fewest payments that clear this bucket: the largest debtor pays the largest creditor and so
  // on, so A pays C directly rather than through B. Deterministic for the same balances.
  suggestions: GroupSettleSuggestion[];
  myBalance: string;
  // The viewer's own position converted for reading at a glance; null when no rate is available, and
  // never what anybody settles.
  myConvertedBalance: string | null;
}

export interface GroupBalances {
  groupId: number;
  // One per currency the group actually has an open position in, alphabetical. A bucket that has been
  // fully settled disappears entirely, so an empty list means everyone is square — or that nothing
  // has been shared yet, which the caller tells apart by whether any expense exists.
  buckets: GroupCurrencyBalance[];
  displayCurrency: string | null;
  skippedCurrencies: string[];
}

export interface GroupSettlement {
  id: number;
  groupId: number;
  fromMemberId: number;
  fromDisplayName: string;
  toMemberId: number;
  toDisplayName: string;
  date: string;
  amount: string;
  currency: string;
  status: SettlementStatus;
  /*
   * The two cash legs, each belonging to one side. Each amount is set only when that side's account
   * was denominated differently from the bucket — within one currency the account moved exactly what
   * the bucket cleared, so a second copy of the figure would be a second thing to keep in step.
   */
  fromAccountId: number | null;
  fromAmount: string | null;
  toAccountId: number | null;
  toAmount: string | null;
  confirmedAt: string | null;
  notes: string | null;
  // Resolved server-side from who the caller is. Render from these rather than re-deriving the rule.
  canConfirm: boolean;
  canDelete: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface GroupMoneySettings {
  groupId: number;
  defaultSplitMethod: SplitMethod;
  autoFinaliseSettlements: boolean;
}

// --- Mappers ---

function mapMemberBalance(raw: GroupMemberBalanceRaw): GroupMemberBalance {
  return {
    memberId: raw.member_id,
    displayName: raw.display_name,
    amount: raw.amount,
    isSelf: raw.is_self,
  };
}

function mapSuggestion(raw: GroupSettleSuggestionRaw): GroupSettleSuggestion {
  return {
    fromMemberId: raw.from_member_id,
    fromDisplayName: raw.from_display_name,
    toMemberId: raw.to_member_id,
    toDisplayName: raw.to_display_name,
    amount: raw.amount,
  };
}

function mapCurrencyBalance(raw: GroupCurrencyBalanceRaw): GroupCurrencyBalance {
  return {
    currency: raw.currency,
    balances: raw.balances.map(mapMemberBalance),
    suggestions: raw.suggestions.map(mapSuggestion),
    myBalance: raw.my_balance,
    myConvertedBalance: raw.my_converted_balance,
  };
}

function mapBalances(raw: GroupBalancesRaw): GroupBalances {
  return {
    groupId: raw.group_id,
    buckets: raw.buckets.map(mapCurrencyBalance),
    displayCurrency: raw.display_currency,
    skippedCurrencies: raw.skipped_currencies,
  };
}

function mapSettlement(raw: GroupSettlementRaw): GroupSettlement {
  return {
    id: raw.id,
    groupId: raw.group_id,
    fromMemberId: raw.from_member_id,
    fromDisplayName: raw.from_display_name,
    toMemberId: raw.to_member_id,
    toDisplayName: raw.to_display_name,
    date: raw.date,
    amount: raw.amount,
    currency: raw.currency,
    status: raw.status,
    fromAccountId: raw.from_account_id,
    fromAmount: raw.from_amount,
    toAccountId: raw.to_account_id,
    toAmount: raw.to_amount,
    confirmedAt: raw.confirmed_at,
    notes: raw.notes,
    canConfirm: raw.can_confirm,
    canDelete: raw.can_delete,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

function mapMoneySettings(raw: GroupMoneySettingsRaw): GroupMoneySettings {
  return {
    groupId: raw.group_id,
    defaultSplitMethod: raw.default_split_method,
    autoFinaliseSettlements: raw.auto_finalise_settlements,
  };
}

// --- API functions ---

/*
 * Every member's position per currency, plus the fewest payments that clear each bucket.
 *
 * `currency` asks for the glance figure beside each bucket only — the buckets themselves are always
 * in their own currency and are never converted, because that is what a settle line has to be.
 *
 * Returns null for a group that does not exist or that the caller is not a member of, so the hub
 * renders notFound() either way and an id cannot be probed.
 */
export async function getGroupBalances(
  groupId: number,
  currency?: string,
): Promise<GroupBalances | null> {
  const query = currency ? `?currency=${encodeURIComponent(currency)}` : '';
  const res = await authenticatedFetch(`/groups/${groupId}/balances${query}`, { method: 'GET' });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error('Failed to fetch group balances');
  return mapBalances(await res.json());
}

// The group's recorded settlements and write-offs, newest first. Null for the same two reasons.
export async function getGroupSettlements(groupId: number): Promise<GroupSettlement[] | null> {
  const res = await authenticatedFetch(`/groups/${groupId}/settlements`, { method: 'GET' });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error('Failed to fetch group settlements');
  const raw: GroupSettlementRaw[] = await res.json();
  return raw.map(mapSettlement);
}

// The money the group holds in common: the split a new expense starts on, and whether a recorded
// settlement is confirmed on the spot. Null for the same two reasons.
export async function getGroupMoneySettings(groupId: number): Promise<GroupMoneySettings | null> {
  const res = await authenticatedFetch(`/groups/${groupId}/money-settings`, { method: 'GET' });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error('Failed to fetch group money settings');
  return mapMoneySettings(await res.json());
}
